### PR TITLE
feat: implement related contracts mechanism (depth=1)

### PR DESCRIPTION
**Pull request by sanity**
## Problem

Related contracts allow a contract's `validate_state` to return `RequestRelated(contract_ids)` requesting other contracts' state for validation. This enables use cases like River room membership gates (contract B validates against contract A's state) and Bitcoin header verification. The mechanism was partially wired up but had broken code paths, no safety limits, no real tests, and a buggy recursive validation flow with `DEPENDENCY_CYCLE_LIMIT_GUARD = 100`.

## Approach

**Depth=1 only, one round.** When `validate_state` returns `RequestRelated(ids)`, we fetch those contracts locally and re-call `validate_state` exactly once. If the second call also returns `RequestRelated`, that's an error. This prevents amplification attacks where a malicious contract requests new contracts on every retry.

Key design decisions:
- **Conservative start**: depth=1 is easy to reason about and increase later
- **One round only**: contracts must declare all dependencies in one call
- **Local fetch only**: the `fetch_related_for_validation` helper uses `lookup_key` + `state_store.get()` (available on all executor types). Network fetch happens automatically via GET auto-subscribe when contracts are first retrieved
- **Network inbound deferred**: when another node broadcasts an update, `RequestRelated` still errors — the sending node already validated. Filing a separate issue for this

Abuse prevention:
- Self-reference explicitly rejected before any fetch
- Max 10 related contracts per request
- 10s overall fetch timeout
- Empty `RequestRelated(vec![])` rejected
- Duplicate IDs deduped via HashSet
- Second `RequestRelated` after providing data → error (depth>1 not supported)

### Changes

1. **`fetch_related_for_validation` helper** — shared validation logic with all safety checks
2. **Simplified `verify_and_store_contract`** — replaced 100-iteration loop with clean depth=1 flow
3. **Fixed 4 validation call sites** — new PUT, post-merge, `perform_contract_put`, `get_updated_state`
4. **Extended `MockWasmRuntime`** — `ValidateOverride` enum enables per-contract validation behavior in tests
5. **Updated `.claude/rules/contracts.md`** — documented depth=1 design decisions

## Testing

10 new unit tests covering all paths and edge cases:

| Test | What it validates |
|------|-------------------|
| `test_put_with_related_succeeds` | PUT gated contract after gate exists locally |
| `test_put_with_gate_closed` | Validation returns Invalid |
| `test_put_self_reference_rejected` | Contract requests own ID → error |
| `test_put_too_many_related_rejected` | >10 related contracts → error |
| `test_put_missing_related_errors` | Non-existent contract → error, not hang |
| `test_depth_exceeded_rejected` | `AlwaysRequestRelated` → depth>1 error |
| `test_empty_request_related` | Empty vec → error |
| `test_duplicate_ids_deduped` | Same ID twice → fetched once, succeeds |
| `test_existing_contract_put_with_related` | Merge path fetches related |
| `test_update_with_related_succeeds` | Delta update path fetches related |

Full test suite: 2108 passed, 0 failures, clippy clean, fmt clean.

[AI-assisted - Claude]

### CODE REVIEW

File: crates/core/src/contract/executor/runtime.rs [hunk 1]
```diff
@@ -3067,129 +3261,58 @@ impl Executor<Runtime> {
...
-                                    trying_state = result.state;
-                                    trying_contract = Some(contract);
-                                    continue;
-                                }
-                            }
-                        }
-                    }
-                    continue;
+        // Validate with depth=1 related contract resolution
+        let result = self
```
Comment (iduartgomez) [inline crates/core/src/contract/executor/runtime.rs:3313]: should document here for future reference to protect against depth requests

Comment (sanity) [reply crates/core/src/contract/executor/runtime.rs:3313]: Added a detailed depth protection comment at this call site documenting all the safety mechanisms: depth=1 limit, one-round-only, self-reference rejection, max 10 contracts, and timeout. Points to the constants for reference.

[AI-assisted - Claude]
